# Homework 2 - Data from the web

### Useful Imports

In [1]:
# Import libraries
%matplotlib inline
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os.path
import sys
import re 

# those two packages must be installed using for example ```pip install editdistance```
import editdistance
import unidecode
sns.set_context('notebook')
pd.options.mode.chained_assignment = None  # default='warn', Mutes warnings when copying a slice from a DataFrame.

## 1 - Top Universities

In particular, extract the following fields for each university: name, rank, country and region, number of faculty members (international and total) and number of students (international and total)

In [2]:
ROOT_URL_TOP_UNIV = "https://www.topuniversities.com/"

In [3]:
def progress(count, total, suffix=''):
    """ Shows the progress of a given action 
    
    @params:
    - count : the current count of done operations
    - total : the total number of operation to do
    - suffix : a message printed after the progress bar
    """
    
    bar_len = 60
    filled_len = int(round(bar_len * count / float(total)))

    percents = round(100.0 * count / float(total), 1)
    bar = '#' * filled_len + '-' * (bar_len - filled_len)

    sys.stdout.write('[%s] %s%s ... %s\r' % (bar, percents, '%', suffix))
    sys.stdout.flush()  # As suggested by Rom Ruben

In [4]:
def get_number(enclosing_class_name,url,det_soup):
    """Tries to extract the number we look for. Will return a NaN 
    if the details page doesn't contain such a number
    
    @params:
    - enclosing_class_name : the class of the div into which the number is contained.
    - url : the suffix url of the school (to print an error if not found)
    - det_soup : the soup of the details page for the given school
    
    """
    try:
        n = det_soup.find("div",class_=enclosing_class_name).find("div",class_="number").text.strip('\n')
    except(AttributeError) as e:
        print("Couldn't find '{}' for {}".format(enclosing_class_name,url))
        n = 'NAN'
    return n

def get_details(url):
    """Finds the detail of a given university on the details page
    
    @params:
    - url : the suffix url of the universirty of interest
    
    """
    details_url = ROOT_URL_TOP_UNIV + url
    det_request = requests.get(details_url)
    det_soup = BeautifulSoup(det_request.text, 'lxml')
    labels = ['total faculty','inter faculty','total student','total inter']

    staff_total,staff_inter,student_total,student_inter = [get_number(label,url,det_soup) for label in labels]
    
    return staff_total,staff_inter,student_total,student_inter

In [5]:
def get_ranking_topUni(path_to_dump):
    """
    Will return a dataframe containing the data of the rankings published on www.topuniversities.com.
    (in order to minimize the internet traffic every time the notebook is ran, the data is serialized using pickle)
    
    @params:
    - path_to_dump : path were the dump of the data should be (will use it if already present and save there otherwise)
    """
    r = requests.get('https://www.topuniversities.com/sites/default/files/qs-rankings-data/357051.txt?_=1508057903494')
    if(os.path.isfile(path_to_dump)):
        return pd.read_pickle(path_to_dump)

    top_200 = []

    for (i,uni_dict) in enumerate(r.json()['data']):
        if(i<200):
            title = uni_dict['title']
            rank = uni_dict['rank_display']
            country = uni_dict['country']
            region = uni_dict['region']
            url = uni_dict['url']

            # In case of tie, the rank is prefixed with an '=', we get rid of it.
            rank = rank.strip('=')

            # We then get the details from the details page
            staff_total,staff_inter,student_total,student_inter = get_details(url)
            progress(i+1,200,'Scraping the info')

            top_200.append({'School Name': title, 
                            'Rank': rank, 
                            'Country': country,
                            'Region': region,
                            'Total Staff': staff_total,
                            'International Staff': staff_inter,
                            'Total Student': student_total,
                            'International Student': student_inter
                           })
            
    df = pd.DataFrame.from_dict(top_200)
    
    
    # We set Rank as an index
    df.set_index('Rank',inplace=True)
    
    # Then we convert the number columns to integers
    cols = ['Total Staff','International Staff','Total Student','International Student']
    df[cols] = df[cols].apply(lambda x: pd.to_numeric(x.astype(str)
                                                       .str.replace(',',''), errors='coerce',downcast='integer'))

    # We return the dataframe ordered as we please
    df = df[['School Name','Total Staff','International Staff','Total Student','International Student','Country','Region']]
    
    # We serialize it using pickle so that we do not have to download it again
    df.to_pickle(path_to_dump)
    return df

In [30]:
df_topUni = get_ranking_topUni('topuniversities_backup')
df_topUni.tail(7)

School Name  Total Staff  \
Rank                                                                   
192                                     Keio University         3905   
195                                Stockholm University         2154   
195                   Universitat Autònoma de Barcelona         2187   
195                                Texas A&M University         3446   
199   Instituto Tecnológico y de Estudios Superiores...         1822   
200                               Maastricht University         1277   
201                                Universidad de Chile         2256   

      International Staff  Total Student  International Student  \
Rank                                                              
192                 261.0          33500                   2178   
195                 646.0          28281                   2636   
195                 230.0          31986                   3848   
195                 206.0          60294                   4900   
199                 821.0          13376                   1412   
200                 502.0          16385                   8234   
201                 127.0          38848                   2134   

            Country         Region  
Rank                                
192           Japan           Asia  
195          Sweden         Europe  
195           Spain         Europe  
195   United States  North America  
199          Mexico  Latin America  
200     Netherlands         Europe  
201           Chile  Latin America

** Remark: there is a mistake in the data,** if we take 200 rows of school we obtain the last school having rank 201. Indeed there is a mistake, From Keio University which is at rank 192, if we count the ranks then "Instituto Tecnologico [...]" should be at rank 198.

### 1.a - Best university in terms of ratio faculty members over students

In order to determine which is the best university regarding ratio between faculty members and students, we create a new column called ```Fac/stud ratio```. Then we sort the rows according to values in this column. (because we were asked for the best universities we display the ranking according to this value)

In [7]:
df_topUni['Fac/stud ratio'] = df_topUni['Total Staff']/df_topUni['Total Student']
sorted_by_facStudRatio = df_topUni.sort_values('Fac/stud ratio',ascending=False)
sorted_by_facStudRatio.head()

School Name  Total Staff  \
Rank                                                              
4     California Institute of Technology (Caltech)          953   
16                                 Yale University         4940   
6                             University of Oxford         6750   
5                          University of Cambridge         5490   
17                        Johns Hopkins University         4462   

      International Staff  Total Student  International Student  \
Rank                                                              
4                   350.0           2255                    647   
16                 1708.0          12402                   2469   
6                  2964.0          19720                   7353   
5                  2278.0          18770                   6699   
17                 1061.0          16146                   4105   

             Country         Region  Fac/stud ratio  
Rank                                                 
4      United States  North America        0.422616  
16     United States  North America        0.398323  
6     United Kingdom         Europe        0.342292  
5     United Kingdom         Europe        0.292488  
17     United States  North America        0.276353

### 1.b - Best university in terms of ratio of international students

We use strategy that is exactly the same than for the 1.a

In [8]:
df_topUni['International ratio'] = df_topUni['International Student']/df_topUni['Total Student']
sorted_by_internationalRatio = df_topUni.sort_values('International ratio',ascending=False)
sorted_by_internationalRatio.head()

School Name  Total Staff  \
Rank                                                                   
35    London School of Economics and Political Scien...         1088   
12      Ecole Polytechnique Fédérale de Lausanne (EPFL)         1695   
8                               Imperial College London         3930   
200                               Maastricht University         1277   
47                           Carnegie Mellon University         1342   

      International Staff  Total Student  International Student  \
Rank                                                              
35                  687.0           9760                   6748   
12                 1300.0          10343                   5896   
8                  2071.0          16090                   8746   
200                 502.0          16385                   8234   
47                  425.0          13356                   6385   

             Country         Region  Fac/stud ratio  International ratio  
Rank                                                                      
35    United Kingdom         Europe        0.111475             0.691393  
12       Switzerland         Europe        0.163879             0.570047  
8     United Kingdom         Europe        0.244251             0.543567  
200      Netherlands         Europe        0.077937             0.502533  
47     United States  North America        0.100479             0.478062

### 1.c - Aggregated by country

Because it wasn't so clear how we should aggregate the data, we decided to consider the ```mean()``` as aggregation function which will give us the ranking of countries according to :
* ratio of faculty staff over students
* ratio of international students

In [9]:
best_countries_by_facStudRatio = df_topUni.groupby('Country').mean().sort_values('Fac/stud ratio',ascending=False)
best_countries_by_facStudRatio[['Fac/stud ratio']].head()

Fac/stud ratio
Country                     
Russia              0.221910
Denmark             0.186580
Saudi Arabia        0.175828
Singapore           0.162279
Japan               0.155840

In [10]:
best_countries_by_internationalRatio = df_topUni.groupby('Country').mean().sort_values('International ratio',ascending=False)
best_countries_by_internationalRatio[['International ratio']].head()

International ratio
Country                            
United Kingdom             0.351308
Australia                  0.346878
Switzerland                0.313816
Hong Kong                  0.312148
Austria                    0.306095

### 1.d - Aggregated by region

Very similar strategy than the one used in 1.c

In [11]:
best_regions_by_facStudRatio = df_topUni.groupby('Region').mean().sort_values('Fac/stud ratio',ascending=False)
best_regions_by_facStudRatio[['Fac/stud ratio']].head()

Fac/stud ratio
Region                       
North America        0.145407
Asia                 0.134673
Europe               0.120003
Latin America        0.096779
Africa               0.088450

In [12]:
best_regions_by_internationalRatio = df_topUni.groupby('Region').mean().sort_values('International ratio',ascending=False)
best_regions_by_internationalRatio[['International ratio']].head()

International ratio
Region                            
Oceania                   0.329077
Europe                    0.245932
North America             0.203583
Africa                    0.169703
Asia                      0.132394

# todo : bar plots for the results

## 2 - Times Higher Education

Because we do not have direct access to some informations (e.g. International Students, Total Staff) we will infer them using the informations we have.

In [13]:
def get_ranking_times(path_to_dump):
    if(os.path.isfile(path_to_dump)):
        return pd.read_pickle(path_to_dump)
    
    r = requests.get('https://www.timeshighereducation.com/sites/default/files/the_data_rankings/world_university_rankings_2018_limit0_369a9045a203e176392b9fb8f8c1cb2a.json')

    top_200 = []
    for (i,uni_dict) in enumerate(r.json()['data']):
        if(i<200):
            title = uni_dict['name']
            # We delete the '=' sign in the case of tie to be consistent with the first one
            rank = uni_dict['rank'].replace('=','')
            country = uni_dict['location']

            # We delete the thousands separator
            student_total = int(uni_dict['stats_number_students'].replace(',',''))
            pc_inter_student = float(uni_dict["stats_pc_intl_students"].replace('%',''))/100
            student_to_staff_ratio = float(uni_dict["stats_student_staff_ratio"])
            
            # infer some informations using the percentage and ratio
            student_inter = student_total*pc_inter_student
            staff_total = student_total/student_to_staff_ratio
            
            

            top_200.append({'School Name': title, 
                                'Rank': rank, 
                                'Country': country,
                                'International Student' : student_inter,
                                'Total Staff' : staff_total,
                                'Total Student': student_total,
                               })
    df = pd.DataFrame.from_dict(top_200)
    df.to_pickle(path_to_dump)
    return df

In [14]:
df_times = get_ranking_times('times_backup')
df_times.head()

Country  International Student Rank  \
0  United Kingdom                7755.42    1   
1  United Kingdom                6436.15    2   
2   United States                 596.43    3   
3   United States                3485.90    3   
4   United States                3800.18    5   

                             School Name  Total Staff  Total Student  
0                   University of Oxford  1822.232143          20409  
1                University of Cambridge  1687.064220          18389  
2     California Institute of Technology   339.846154           2209  
3                    Stanford University  2112.666667          15845  
4  Massachusetts Institute of Technology  1284.712644          11177

## 3 - Merging the two data frames
### 3.1 - Matching the school names

Because we do not have the same information than in the first table we try to get the informations using the first table. But first we need to see how the school names can differ between the two tables (e.g. "Massachusetts Institute of Technology (MIT)" is different from "Massachusetts Institute of Technology"). 

We will proceed in two steps: 
1. We will first clean the school names in both dataframes by :
    * Converting to Lowercase
    * Deleting accents
    * Deleting ponctuation {,.-}
    * Deleting the initials or the school put in parenthesis (e.g. (EPFL) or (MIT))
    * Deleting the trailing "the"
2. We will try to identify similar names by :
    * Considering if they belong to the same country
    * Seeing if one is contained in the other
    * Looking at the [Levenstein Distance](https://en.wikipedia.org/wiki/Levenshtein_distance) between the words that are different between the two names. This distance is considered relative to the maximum length of both names to have a relative number for all school names. 

In [15]:
def how_different(df_times,df_topUni):
    """ 
    Will return the intersection set of school names from the two dataframes. 
    This helps us count the number of unmatched school names between the two df.
    
    @params:
    - df_times, df_topUni : the two dataframes for which we wish to compare the 'School Name' values
    
    """
    names_times = set(sorted(list(df_times['School Name'].values)))
    names_topUni = set(sorted(list(df_topUni['School Name'].values)))
    difference = names_times.symmetric_difference(names_topUni)
    return difference

In [16]:
def clean_data(x):
    """
    Performs very basic string cleaning on a serie of School names.
    It puts all names in lower case, gets rid of the accents,
    deletes the symbols that are added between parenthesis (like EPFL),
    the trailing "the" and some ponctuation.
    
    @params:
    - x: the data frame to clean
    """
    # get small latters
    x = x.lower()
    
    # delete accents
    x = unidecode.unidecode(x)
    
    # delete parenthesis, points, comas, dash and trailing "the"
    x = re.sub("[.,]","",x)
    x = re.sub(r"(\s)\-(\s)", r"\1\2",x) #dash between two spaces replace by nothing
    x = re.sub(r"([a-z])\-([a-z])", r"\1 \2",x) #dash between two words replace by space
    x = re.sub("[\(].*?[\)]", "", x)
    x = re.sub("^(the)\s", "",x)
    
    # delete any space before or after string
    return x.strip()

print("Before data cleaning, we have {} different school names.".format(len(how_different(df_times,df_topUni))))
df_times['School Name'] = df_times['School Name'].apply(clean_data)
df_topUni['School Name'] = df_topUni['School Name'].apply(clean_data)
print("After data cleaning, we have {} different school names.".format(len(how_different(df_times,df_topUni))))

Before data cleaning, we have 190 different school names.
After data cleaning, we have 116 different school names.


In [31]:
def get_difference(s1,s2):
    """ 
    Many name of school contain same words "technology","university", 
    we want to get rid of them such that the relative difference 
    really characterises words that are different. 
    Therefore we return only the words for s1 and s2 that are not in both strings.
    
    @params:
    - s1,s2 : the two strings that we wish to compare
    
    """
    # Many name of school contain same words "technology","university", 
    # we want to get rid of them such that the relative difference 
    # really characterises words that are different
    d1 = [w for w in s1.split() if w not in s2]
    d2 = [w for w in s2.split() if w not in s1]
    return " ".join(d1)," ".join(d2)

def check_similarity(s1,s2,threshold,combined):
    """
    This utility function helps us checking in what magnitude are two school names identical.
    It uses 3 checks :
    - are the two school names from the same country
    - is one of the name included in the second
    - what is the Levenshtein distance relative to the max length of the two
    
    @params:
    - s1,s2 : the two school names
    - threshold : the limit for the relative Levenshtein distance s.t. the names are considered similar
    - combined : a dataframe containing all 'School Name' and 'Country'
    """
    
    o,n = get_difference(s1,s2)
    same_country = (combined.loc[s1,'Country'] == combined.loc[s1,'Country'])
    
    if(len(o) == 0 or len(n)==0):
        # One school name was contained in the other
        return same_country
    else:
        max_length =  len(n) if (len(n) > len(o)) else len(o)
        relative_diff = float(editdistance.eval(o,n))/max_length

        # Then we compute some of our conditions that determine if the strings are similar or not.
        similar_ortho = (relative_diff < threshold)

        return (similar_ortho and same_country)

def get_potentially_similar(df_times,df_topUni):
    """
    Utility function returning a mapping of the School names 
    in df_times to School names in df_topUni that are considered similar
    
    @params:
    - df_times,df_topUni: the two dataframes between which we wish to match the School Names
    """
    reduced_times = df_times.set_index('School Name')[['Country']].reset_index()
    reduced_topUni = df_topUni.set_index('School Name')[['Country']].reset_index()
    
    # We want a dataframe giving us the country for each school
    combined = reduced_times.merge(reduced_topUni,how='outer').set_index('School Name')
    
    names_times = set(sorted(list(df_times['School Name'].values)))
    names_topUni = set(sorted(list(df_topUni['School Name'].values)))
    
    unmatched_times = list(names_times-names_topUni)
    unmatched_topUni = list(names_topUni - names_times)
    
    matching = {}
    for name_1 in unmatched_times:
        for name_2 in unmatched_topUni:
            if(name_1 != name_2):
                # we check different conditions
                if(check_similarity(name_1,name_2,0.35,combined)):
                    # They are similar, we will set name_1 to be replaced by name_2
                    matching[name_1] = name_2
    return matching

def make_homogeneous(x,to_homogeneous):
    """
    Returns either the value in to_homogeneous corresponding 
    to x or x itself if it isn't part of the dictionnary.
    
    @params:
    - x : a school name
    - to_homogeneous : the dictionnary of mappings between 
    school names (see get_potentially_similar)
    """
    if(x in to_homogeneous.keys()): return to_homogeneous[x]
    else: return x

def match_school_names(df_times,df_topUni):
    """
    Will modify the school names of df_times to match them when 
    possible with the school names of df_topUni.
    
    @params:
    - df_times, df_topUni: the dataframes of the two rankings 
    for which we wish to match the school names.
    """
    matching = get_potentially_similar(df_times,df_topUni)

    df_times['School Name'] = df_times['School Name'].apply(make_homogeneous,to_homogeneous = matching)

In [18]:
print("Before matching names, we have {} different school names.".format(len(how_different(df_times,df_topUni))))
match_school_names(df_times,df_topUni)
print("After matching names, we have {} different school names.".format(len(how_different(df_times,df_topUni))))

Before matching names, we have 116 different school names.
{'eth zurich  swiss federal institute of technology zurich': 'eth zurich  swiss federal institute of technology', 'humboldt university of berlin': 'humboldt universitat zu berlin', 'karlsruhe institute of technology': 'kit karlsruhe institute of technology', 'technical university of berlin': 'technische universitat berlin', 'university of north carolina at chapel hill': 'university of north carolina chapel hill', 'free university of berlin': 'freie universitaet berlin', 'university of arizona': 'universitat de barcelona', 'trinity college dublin': 'trinity college dublin the university of dublin', "scuola superiore sant'anna": "scuola superiore sant'anna pisa di studi universitari e di perfezionamento", 'wageningen university & research': 'wageningen university', 'university of montreal': 'universite de montreal'}
After matching names, we have 94 different school names.


### 3.2 - Merging

Now that we have only so few different names we can consider that the rest are really different from the two rankings. Indeed we have performed all the matching that we could. We merge the two dataframes based on 'School Name' (using a outer join to keep the names in both tables), we add suffixes to columns of both frames to be able to combine them later on.

In [19]:
df_times_to_merge = df_times
df_topUni_to_merge = df_topUni.reset_index()
merged_df = df_topUni_to_merge.merge(df_times_to_merge,on='School Name',how='outer', suffixes=('_topUni', '_times'))
merged_df.head()

Rank_topUni                            School Name  Total Staff_topUni  \
0           1  massachusetts institute of technology              2982.0   
1           2                    stanford university              4285.0   
2           3                     harvard university              4350.0   
3           4     california institute of technology               953.0   
4           5                university of cambridge              5490.0   

   International Staff  Total Student_topUni  International Student_topUni  \
0               1679.0               11067.0                        3717.0   
1               2042.0               15878.0                        3611.0   
2               1311.0               22429.0                        5266.0   
3                350.0                2255.0                         647.0   
4               2278.0               18770.0                        6699.0   

   Country_topUni         Region  Fac/stud ratio  International ratio  \
0   United States  North America        0.269450             0.335863   
1   United States  North America        0.269870             0.227422   
2   United States  North America        0.193945             0.234785   
3   United States  North America        0.422616             0.286918   
4  United Kingdom         Europe        0.292488             0.356899   

    Country_times  International Student_times Rank_times  Total Staff_times  \
0   United States                      3800.18          5        1284.712644   
1   United States                      3485.90          3        2112.666667   
2   United States                      5284.76          6        2283.820225   
3   United States                       596.43          3         339.846154   
4  United Kingdom                      6436.15          2        1687.064220   

   Total Student_times  
0              11177.0  
1              15845.0  
2              20326.0  
3               2209.0  
4              18389.0

Now because we don't know what ranking to believe in terms of :
* International Students
* Total Students
* Total Staff

We decided to take the mean of the values given by both in the combined dataframe. If we were to trust more one ranking than the other we could use a weighted average.

In [20]:
def avg_col(x,name_col_1,name_col_2):
    """
    Will return the average of two columns of a dataframe ignoring NaN values.
    
    @params:
    - x : dataframe in which the columns are.
    - name_col_1,name_col_2 : the name of the two columns we wish to aggregate.
    """
    if pd.isnull(x[name_col_1]) and pd.isnull(x[name_col_2]):
        print("Houston we have a problem !")
    if pd.isnull(x[name_col_1]):
        return x[name_col_2]
    elif pd.isnull(x[name_col_2]):
        return x[name_col_1]
    else: return (x[name_col_1]+x[name_col_2])/2

In [21]:
merged_df['Total Staff'] = merged_df.apply(avg_col,name_col_1 ='Total Staff_topUni',name_col_2='Total Staff_times',axis=1)
merged_df['Total Student'] = merged_df.apply(avg_col,name_col_1 ='Total Student_topUni',name_col_2='Total Student_times',axis=1)
merged_df['International Student'] = merged_df.apply(avg_col,name_col_1 ='International Student_topUni',name_col_2='International Student_times',axis=1)

In [22]:
merged_df = merged_df[['Rank_topUni','Rank_times','Total Staff','International Staff','Total Student','International Student','Country_times','Region','School Name']]
merged_df = merged_df.set_index('School Name')

In [25]:
merged_df

Rank_topUni Rank_times  \
School Name                                                                
massachusetts institute of technology                       1          5   
stanford university                                         2          3   
harvard university                                          3          6   
california institute of technology                          4          3   
university of cambridge                                     5          2   
university of oxford                                        6          1   
ucl                                                         7        NaN   
imperial college london                                     8          8   
university of chicago                                       9          9   
eth zurich  swiss federal institute of technology          10         10   
nanyang technological university singapore                 11         52   
ecole polytechnique federale de lausanne                   12         38   
princeton university                                       13          7   
cornell university                                         14         19   
national university of singapore                           15         22   
yale university                                            16         12   
johns hopkins university                                   17         13   
columbia university                                        18         14   
university of pennsylvania                                 19         10   
australian national university                             20         48   
duke university                                            21         17   
university of michigan                                     21         21   
king's college london                                      23         36   
university of edinburgh                                    23         27   
tsinghua university                                        25         30   
university of hong kong                                    26         40   
university of california berkeley                          27         18   
northwestern university                                    28         20   
university of tokyo                                        28         46   
hong kong university of science and technology             30         44   
...                                                       ...        ...   
pompeu fabra university                                   NaN        140   
university of cologne                                     NaN        145   
autonomous university of barcelona                        NaN        147   
university of sussex                                      NaN        147   
university of notre dame                                  NaN        150   
tu dresden                                                NaN        155   
ulm university                                            NaN        155   
case western reserve university                           NaN        158   
university of leicester                                   NaN        159   
university of california santa cruz                       NaN        162   
university of erlangen nuremberg                          NaN        162   
vrije universiteit amsterdam                              NaN        165   
university of wurzburg                                    NaN        165   
university of alabama at birmingham                       NaN        168   
tufts university                                          NaN        169   
rutgers the state university of new jersey                NaN        172   
university of munster                                     NaN        173   
universite libre de bruxelles                             NaN        175   
university of luxembourg                                  NaN        179   
paris sud university                                      NaN        181   
university of miami   